msal/src/Microsoft.Identity.Client/Internal/Requests/DeviceCodeRequest.cs

```diff
@@ -0,0 +1,125 @@
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Identity.Core.Helpers;
+using Microsoft.Identity.Core.OAuth2;
+
+namespace Microsoft.Identity.Client.Internal.Requests
+{
+    // todo: 
+    // * unit tests
+    // * integration/end2end tests in internal vsts repo
+    // * sample app / test dev app as a console application for validation
+    // * do we have lab accounts/environments setup for device auth?
+    // * do we need to set SaveToCache in the constructor?
+    internal class DeviceCodeRequest : RequestBase
+    {
+        private readonly Action<DeviceCodeResult> _deviceCodeResultCallback;
+        private DeviceCodeResult _deviceCodeResult;
+
+        public DeviceCodeRequest(AuthenticationRequestParameters authenticationRequestParameters, Action<DeviceCodeResult> deviceCodeResultCallback) 
+            : base(authenticationRequestParameters)
+        {
+            _deviceCodeResultCallback = deviceCodeResultCallback;
+            
+            LoadFromCache = false;  // no cache lookup for token
+            SupportADFS = false;
+        }
+
+        internal override async Task PreTokenRequestAsync()
+        {
+            await base.PreTokenRequestAsync().ConfigureAwait(false);
+
+            OAuth2Client client = new OAuth2Client();
+
+            var deviceCodeScopes = new SortedSet<string>();
+            deviceCodeScopes.UnionWith(AuthenticationRequestParameters.Scope);
+            deviceCodeScopes.Add("offline_access");
+            deviceCodeScopes.Add("profile");
+            deviceCodeScopes.Add("openid");
+
+            client.AddBodyParameter(OAuth2Parameter.ClientId, AuthenticationRequestParameters.ClientId);
+            client.AddBodyParameter(OAuth2Parameter.Scope, deviceCodeScopes.AsSingleString());
+
+            // todo: THIS IS A MAJOR HACK.  Work with Shiung/Henrik on proper way to determine the device code endpoint
+            string deviceCodeEndpoint = AuthenticationRequestParameters.Authority.TokenEndpoint
+                .Replace("token", "devicecode")
+                .Replace("common", "organizations");

```

Comment:
In Integrated Windows Auth you get a pretty good error message when the user tries to use /common - is it not the case here?

Reply:
current recommendations from evo is that we do not use common. the error message returned is not guaranteed to be brilliant and is currently subject to change.